### PR TITLE
Add "none" distribution

### PR DESCRIPTION
**Pull request by snickl**
So here's a first stab.
I had to make the efi symlink conditional because my testing skeleton tree is for an arm system without EFI.
Also I just took the architecture list from Fedora for now, is there a way to say "all of them / don't care"?

### CODE REVIEW

File: mkosi/distributions/none.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+
+class NoneInstaller(DistributionInstaller):
+    @classmethod
+    def filesystem(cls) -> str:
+        return "ext4"
+
+    @classmethod
+    def package_type(cls) -> PackageType:
+        return PackageType.rpm
```
Comment (behrmann) [inline mkosi/distributions/none.py:19]: That seems an odd choice? Maybe we should have an associated package type `none`, too.

Comment (DaanDeMeyer) [inline mkosi/distributions/none.py]: Let's add `none` to `PackageType` and use it here

File: mkosi/distributions/none.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            Architecture.ppc64_le  : "ppc64le",
+            Architecture.riscv64   : "riscv64",
+            Architecture.s390x     : "s390x",
+            Architecture.x86_64    : "x86_64",
+        }.get(arch)
+
+        if not a:
+            die(f"Architecture {a} is not supported")
+
+        return a
```
Comment (behrmann) [inline mkosi/distributions/none.py:66]: I see that it's taken from Fedora, by without making further any reason not do just do
```suggestion
        return str(arch)
```

**Comment by DaanDeMeyer** [comment]
You can rebase on top of https://github.com/systemd/mkosi/pull/1933 once it's merged and it will make this a bit easier, you'll have to implement fewer methods.

So this should also come with a few extra checks. e.g. after running the prepare scripts we should check that `state.root` was actually populated, which I figure we can check by making sure a /usr/lib/os-release file exists.

File: mkosi/resources/mkosi.md [hunk 3]
```diff
@@ -375,7 +375,7 @@ boolean argument: either `1`, `yes`, or `true` to enable, or `0`, `no`,
 
 : The distribution to install in the image. Takes one of the following
   arguments: `fedora`, `debian`, `ubuntu`, `arch`, `opensuse`, `mageia`,
-  `centos`, `openmandriva`, `rocky`, `alma`. If not
+  `centos`, `openmandriva`, `rocky`, `alma`, `none`. If not
```
Comment (DaanDeMeyer) [inline mkosi/resources/mkosi.md:378]: We should document here that if the distribution is `none` that it's up to the user to populate the rootfs via a combination of base trees, skeleton trees and prepare scripts.

File: mkosi/distributions/none.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    def setup(cls, state: MkosiState) -> None:
+        pass
+
+    @classmethod
+    def install(cls, state: MkosiState) -> None:
+        pass
+
+    @classmethod
+    def install_packages(cls, state: MkosiState, packages: Sequence[str], apivfs: bool = True) -> None:
+        pass
```
Comment (DaanDeMeyer) [inline mkosi/distributions/none.py:42]: We should fail explicitly here if `packages` is not empty

File: mkosi/distributions/none.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    @classmethod
+    def remove_packages(cls, state: MkosiState, packages: Sequence[str]) -> None:
+        pass
```
Comment (DaanDeMeyer) [inline mkosi/distributions/none.py:46]: Same here, we should fail if `packages` is not empty

File: mkosi/distributions/__init__.py [hunk 6]
```diff
@@ -81,6 +83,7 @@ def is_centos_variant(self) -> bool:
 
     def is_dnf_distribution(self) -> bool:
         return self in (
+            Distribution.none,
```
Comment (DaanDeMeyer) [inline mkosi/distributions/__init__.py:86]: This should be dropped

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

File: mkosi/distributions/none.py (hunk 1 above)
Comment (snickl) [reply mkosi/distributions/none.py]: Just an oversight. I also added "none", but without testing on my side so far.
#1933 also adds this, so should sort out itself.

Comment (snickl) [reply mkosi/distributions/none.py]: Done

**Comment by DaanDeMeyer** [comment]
We should also add a check in `build_initrd()` that fails if the distribution is `none` since we can't build a default initrd if that's the case.

**Comment by DaanDeMeyer** [comment]
#1933 is merged now so you can rebase on main

**Comment by DaanDeMeyer** [comment]
Please rebase on the latest changes from the main branch and remove the implementations of methods that have default implementation on the main branch

File: mkosi/__init__.py [hunk 7]
```diff
@@ -138,7 +138,8 @@ def install_distribution(state: MkosiState) -> None:
             # Some distributions install EFI binaries directly to /boot/efi. Let's redirect them to /efi
             # instead.
             rmtree(state.root / "boot/efi")
-            (state.root / "boot/efi").symlink_to("../efi")
+            if (state.root / "boot/efi").exists():
```
[the remark this answers is not in the document]
Comment (DaanDeMeyer) [reply mkosi/__init__.py:141]: This still needs to be fixed

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Still missing the requested changes to build_initrd() and the check to see if $BUILDROOT was populated after running the prepare scripts. See previous comments

File: mkosi/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (snickl) [reply mkosi/__init__.py]: I've added the mkdir, but I don't really get it

File: mkosi/resources/mkosi.md (hunk 3 above)
Comment (snickl) [reply mkosi/resources/mkosi.md]: Done, please check the phrasing

**Comment by snickl** [comment]
> We should also add a check in `build_initrd()` that fails if the distribution is `none` since we can't build a default initrd if that's the case.

Done but not sure how to trigger this code path.

Why is this not possible? Just because of the --package cmdline elements?

**Comment by snickl** [comment]
> So this should also come with a few extra checks. e.g. after running the prepare scripts we should check that `state.root` was actually populated, which I figure we can check by making sure a /usr/lib/os-release file exists.

Where would I put this? Before run_build_scripts() in build_image()?

File: mkosi/__init__.py [hunk 8]
```diff
@@ -969,6 +971,9 @@ def build_initrd(state: MkosiState) -> Path:
     if symlink.exists():
         return symlink.resolve()
 
+    if state.config.distribution == Distribution.none:
+        die(f"Building an initrd is not supported with distribution {state.config.distribution}")
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:975]: "Building a default initrd ..."

File: mkosi/distributions/none.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    @classmethod
+    def setup(cls, state: MkosiState) -> None:
+        pass
+
+    @classmethod
+    def install(cls, state: MkosiState) -> None:
+        pass
+
+    @classmethod
+    def install_packages(cls, state: MkosiState, packages: Sequence[str], apivfs: bool = True) -> None:
```
Comment (DaanDeMeyer) [inline mkosi/distributions/none.py:20]: The `apivfs` argument can be dropped

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Almost good to go

> Where would I put this? Before run_build_scripts() in build_image()?

Never mind, we'll merge without that check, I'll add it myself in a follow up.

> Why is this not possible? Just because of the --package cmdline elements?

Yes, our default initrd should just consist out of systemd, udev and kmod, but there's no way for us to request just those packages to be installed when there's no distribution, so the initrd has to be provided explicitly by the user.

File: mkosi/__init__.py [hunk 10]
```diff
@@ -176,6 +178,18 @@ def remove_packages(state: MkosiState) -> None:
             die(f"Removing packages is not supported for {state.config.distribution}")
 
 
+def check_none_root_populated(state: MkosiState) -> None:
+    """If no distribution was installed, check that root was populated by looking for os-release file."""
+    if state.config.distribution != Distribution.none:
+        return
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:184]: This shouldn't be specific to the `none` distribution, we should do this for every distribution.

File: mkosi/__init__.py [hunk 11]
```diff
@@ -176,6 +178,18 @@ def remove_packages(state: MkosiState) -> None:
...
+
+    osrelease = state.root / "usr/lib/os-release"
+    if osrelease.is_file():
+        return
+
+    die("Root must be populated from skeleton tree or by other means for distribution 'none'")
```
Comment (keszybz) [inline mkosi/__init__.py:190]: I think it'd be better to say:
```python
die(f"{osrelease} is not a file.",
    hint=""Root must be populated from skeleton tree or by other means for distribution 'none'")
```

**Review by keszybz — COMMENTED** [review]
This seems reasonable.

File: mkosi/__init__.py [hunk 12]
```diff
@@ -176,6 +178,14 @@ def remove_packages(state: MkosiState) -> None:
             die(f"Removing packages is not supported for {state.config.distribution}")
 
 
+def check_root_populated(state: MkosiState) -> None:
+    """Check that the root was populated by looking for a os-release file."""
+    osrelease = state.root / "usr/lib/os-release"
+    if not osrelease.is_file():
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:184]: I would just use `exists()` here

File: mkosi/__init__.py [hunk 13]
```diff
@@ -176,6 +178,14 @@ def remove_packages(state: MkosiState) -> None:
...
+        die(f"{osrelease} is not a file.",
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:185]: "is not a file" => "not found"

File: mkosi/resources/mkosi.md [hunk 14]
```diff
@@ -395,8 +395,11 @@ boolean argument: either `1`, `yes`, or `true` to enable, or `0`, `no`,
 
 : The distribution to install in the image. Takes one of the following
   arguments: `fedora`, `debian`, `ubuntu`, `arch`, `opensuse`, `mageia`,
-  `centos`, `rhel-ubi`, `openmandriva`, `rocky`, `alma`. If not
-  specified, defaults to the distribution of the host.
+  `centos`, `rhel-ubi`, `openmandriva`, `rocky`, `alma`, `none`.
```
Comment (DaanDeMeyer) [inline mkosi/resources/mkosi.md:398]: Can you update the `Supported Distributions` section as well?

File: mkosi/__init__.py (hunk 11 above)
Comment (snickl) [reply mkosi/__init__.py]: Thanks, combined with the request to make it distribution-agnostic

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Can you also make `config_default_distribution()` return `Distribution.none` if no distribution can be found?

File: mkosi/__init__.py (hunk 10 above)
Comment (snickl) [reply mkosi/__init__.py]: Done

File: mkosi/distributions/none.py (hunk 9 above)
Comment (snickl) [reply mkosi/distributions/none.py]: Done

File: mkosi/__init__.py (hunk 8 above)
Comment (snickl) [reply mkosi/__init__.py]: Done

**Comment by keszybz** [comment]
Alsa unmark it as draft.

File: mkosi/__init__.py (hunk 13 above)
Comment (snickl) [reply mkosi/__init__.py]: Done

File: mkosi/resources/mkosi.md (hunk 14 above)
Comment (snickl) [reply mkosi/resources/mkosi.md:398]: Haven't seen that section before. I removed the explanation from --distribution down to that section and expanded it a little bit.

File: mkosi/__init__.py (hunk 12 above)
Comment (snickl) [reply mkosi/__init__.py]: Done

**Comment by keszybz** [comment]
LGTM.

**Review by keszybz — APPROVED** [review]

**Pull request merged**